test/units/module_utils/test_api.py

```diff
@@ -46,3 +46,119 @@ def login_database():
 
         with pytest.raises(Exception):
             login_database()
+
+    def test_no_retries(self):
+
+        @retry()
+        def login_database():
+            assert False, "Should not execute"
+
+        login_database()
+
+
+class TestRetryWithDelaysAndCondition:
+
+    def test_empty_retry_iterator(self):
+        self.counter = 0
+
+        @retry_with_delays_and_condition(backoff_iterator=[])
+        def login_database():
+            self.counter += 1
+
```

Comment:
Maybe we could avoid relying a shared instance attribute?
```suggestion
        def login_database():
            login_database.counter += 1
        login_database.counter = 0

```

Reply:
Another (and possibly cleaner) way to do this is to use `pytest-mock`:

```python
class SpiedOnTarget:
    @staticmethod
    def do_a_thing():
        return None

def test_empty_retry_iterator(mocker):
    decorate_with_no_retries = retry_with_delays_and_condition(backoff_iterator=[])
    spyable = mocker.spy(SpiedOnTarget, do_a_thing)
    invoke_retriable = decorate_with_no_retries(SpiedOnTarget.do_a_thing)

    invoke_retriable()
    assert spyable.call_count == 0
```

Reply:
`mocker.spy` is great. I used your first suggestion though, since the retry-able functions that need an initial failure are actually using the call count.

Reply:
Oh, the other tests?